source/digits_hits/include/GateRestrictedVolume.hh

```diff
@@ -0,0 +1,71 @@
+/** ----------------------
+  Copyright (C): OpenGATE Collaboration
+  This software is distributed under the terms
+  of the GNU Lesser General  Public Licence (LGPL)
+  See LICENSE.md for further details
+  ----------------------*/
+
+
+#ifndef GateRestrictedVolume_h
+#define GateRestrictedVolume_h 1
+
+#include "globals.hh"
+#include <iostream>
+
+#include "GateVPulseProcessor.hh"
+
+class GateRestrictedVolumeMessenger;
+
+
+/*! \class  GateRestrictedVolume
+    \brief  Pulse-processor modelling a simple volume discriminator.
+
+    - The method ProcessOnePulse of this class models a simple
+      volume discriminator: any input pulse whose volume name is not as
+      defined is copied into the output pulse-list.
+      On the contrary, any input pulse whose name is the same as 
+      specified is discarded.
+
+      \sa GateVPulseProcessor
+*/
+class GateRestrictedVolume : public GateVPulseProcessor
+{
+  public:
+
+    //! Constructs a new module attached to a GateDigitizer
+    GateRestrictedVolume(GatePulseProcessorChain* itsChain, const G4String& itsName);
+    //! Destructor
+    virtual ~GateRestrictedVolume();
+
+    //! Returns the restricted volume
+    G4String GetRestrictedVolume() { return m_restrictedVolume; }
+
+    //! Set the trestricted volume
+    void SetRestrictedVolume(G4String val) { m_restrictedVolume = val; }
+
+    //! Returns the restricted volume
+    G4String GetRestrictionMode() { return m_restrictionMode; }
+
+    //! Set the restriction mode
+    void SetRestrictionMode(G4String val) { m_restrictionMode = val; }
+
+
+    //! Implementation of the pure virtual method declared by the base class GateClockDependent
+    //! print-out the attributes specific of the module
+    virtual void DescribeMyself(size_t indent);
+    
+  protected:
+    //! Implementation of the pure virtual method declared by the base class GateVPulseProcessor
+    //! This methods processes one input-pulse
+    //! It is is called by ProcessPulseList() for each of the input pulses
+    //! The result of the pulse-processing is incorporated into the output pulse-list
+    void ProcessOnePulse(const GatePulse* inputPulse,GatePulseList&  outputPulseList);
+
+  private:
+    G4String m_restrictedVolume;                             //!< Rejected volume
+    G4String m_restrictionMode;
+    GateRestrictedVolumeMessenger *m_messenger = nullptr;    //!< Messenger
```

Comment:
You can replace this with unique pointer:
`std::unique_ptr<GateRestrictedVolumeMessenger> m_messenger(new GateRestrictedVolumeMessenger(this));`

With this you dont have to call delete.

Dont forget to add:
`#include<memory>`

Reply:
I tried to implement as you indicated but got an error:
`/home/jakub/Dev/git/Gate/source/digits_hits/include/GateRestrictedVolume.hh:72:65: error: expected identifier before ‘new’
   72 |     std::unique_ptr<GateRestrictedVolumeMessenger> m_messenger (new GateRestrictedVolumeMessenger(this)); //!< Messenger
      |                                                                 ^~~
/home/jakub/Dev/git/Gate/source/digits_hits/include/GateRestrictedVolume.hh:72:65: error: expected ‘,’ or ‘...’ before ‘new’
make[2]: *** [CMakeFiles/GateLib.dir/build.make:2923: CMakeFiles/GateLib.dir/source/digits_hits/src/GateRestrictedVolume.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:476: CMakeFiles/GateLib.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
`

Reply:
I guess because this is not defined yet. You can try to use `std::unique_ptr<GateRestrictedVolumeMessenger> m_messenger;` and later use make_unique in the constructor

Reply:
I found that by default GATE doesn't support C++14 of which make_uinique is member of the std. I had to add few lines to CMakeLists.txt to make it work.

Reply:
If we want to upstream this change, changing CMakeLists is not the way to go (as I think Gate won't accept this change).

If you are only missing support for make_unique, please do not use it and just create unique ptr yourself (you do not need to use ``make_unique`` function).

Reply:
Unique pointer issue has been already solved